openedx/adg/lms/applications/test/test_views.py

```diff
@@ -282,3 +283,187 @@ def test_get_no_user_application_exists_for_application_success_view(get_request
 
     response = ApplicationSuccessView.as_view()(get_request_for_application_success_view)
     assert response.status_code == 400
+
+
+# ------- Application Cover Letter View tests below -------
+
+
+@pytest.fixture(name='cover_letter_view_get_request')
+def cover_letter_view_get_request_fixture(request_factory, user):
+    """
+    Return a HttpRequest object for cover letter get request
+
+    Args:
+        request_factory (RequestFactory): factory to make requests
+        user (User): The user that is logged in
+
+    Returns:
+        HttpRequest object
+    """
+    request = request_factory.get(reverse('application_cover_letter'))
+    request.user = user
+    return request
+
+
+@pytest.fixture(name='cover_letter_view_post_request')
+def cover_letter_view_post_request_fixture(request_factory, user):
+    """
+    Return a HttpRequest object for cover letter post request
+
+    Args:
+        request_factory (RequestFactory): factory to make requests
+        user (User): The user that is logged in
+
+    Returns:
+        HttpRequest object
+    """
+    request = request_factory.post(reverse('application_cover_letter'))
+    request.user = user
+    return request
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('is_get_request', [True, False], ids=['get_request', 'post_request'])
+def test_redirection_of_a_user_without_login_for_cover_letter_view(is_get_request):
+    """
+    Test if an unauthenticated user is redirected to login page on sending a get or post request.
+    """
+    if is_get_request:
+        response = Client().get(reverse('application_cover_letter'))
+    else:
+        response = Client().post(reverse('application_cover_letter'))
+
+    assert COVER_LETTER_REDIRECT_LOGIN_URL in response.url
+
+
+@pytest.mark.django_db
+def test_get_request_of_user_with_complete_application_cover_letter_view(cover_letter_view_get_request):
+    """
+    Test that if a user who has already completed written application sends a get request, they are redirected to the
+    Application Hub page.
+    """
+    cover_letter_view_get_request.user.application_hub.set_is_written_application_completed()
+
+    response = CoverLetterView.as_view()(cover_letter_view_get_request)
+
+    assert response.get('Location') == reverse('application_hub')
+    assert response.status_code == 302
+
+
+@pytest.mark.django_db
+def test_post_request_of_user_with_complete_application_cover_letter_view(cover_letter_view_post_request):
+    """
+    Test that if a user who has already completed written application sends a post request, http 400 is returned.
+    """
+    cover_letter_view_post_request.user.application_hub.set_is_written_application_completed()
+
+    response = CoverLetterView.as_view()(cover_letter_view_post_request)
+
+    assert response.status_code == 400
```

Comment:
After merging tests, the code looks lot more clean and readable. For further improvement and consistency, can we also merge these two tests. We can receive `cover_letter_view_get_request` and `cover_letter_view_post_request` as different parameters, in the same test.